plugins/imagebase/plugin.js

```diff
@@ -554,9 +560,11 @@
 						focused._refreshCaption( sender );
 					}
 
-					if ( previous && hasWidgetFeature( previous, 'caption' ) ) {
-						previous._refreshCaption( sender );
-					}
+					activeWidgets.forEach( function( item ) {
+						if ( item && hasWidgetFeature( item, 'caption' ) ) {
+							item._refreshCaption( sender );
+						}
+					} );
```

Comment:
It seems that the intention of previous code is to have currently focused (`focused`) and previously focused (`previous`) widget reference to update its captions which makes sense (as you are not able to focus more than one widget having last two is sufficient). However, getting `previous` instance based on `'figcaption[data-cke-caption-active]'` does not work.

I see 3 solutions here:

* One proposed above (the only problem is that it will refresh captions of all widgets with active captions not only those recently focused/selected).
* Stay with `previous` approach, but take the variable out of the scope of listener (so it will be defined inside `setUp` function together with `listeners`). Last thing happening in the listener should be updating it like `previous = focused;` so next time listener is called, `previous` will really point to previously focused widget.
* Third option is getting `previous` like `previous = editor.widgets.getByElement( editor.editable().findOne( 'figcaption.cke_widget_editable_focused[data-cke-caption-active]' ) );`. It is based on the fact that above listener has `9` priority and [`selectionChange` one in widgets which updates `cke_widget_editable_focused` has `10` priority](https://github.com/ckeditor/ckeditor-dev/blob/7a1b01f3cbd7a0ad7c41c3f20a09b5e6de15810b/plugins/widget/plugin.js#L2746) (lower). So when above listener is called, previously focused widget will still have `cke_widget_editable_focused` class and can be obtained this way.

Using `previous` approach seems a little smarter (and more efficient), but I am not sure about it. Asking for opinion, WDYT @mlewand @engineering-this ?

Reply:
Quite a few propositions @f1ames 😉

> One proposed above (the only problem is that it will refresh captions of all widgets with active captions not only those recently focused/selected).

I'd rather avoid iterating over all the widgets, it's unnecessary and puts a performance footprint on CKE where we can avoid it.

> Stay with previous approach (…) previously focused widget.

It smells pretty much like a global variable :D However since it would be declared in `setup` it would be one variable per editor, which should do a good job keeping track of last, actual caption. So I think there's a potential in this idea.

Maybe with that we would be able to get rid of `[data-cke-caption-active]` attributes all along?

> Third option is getting previous like (…)

I see - it would, yes - but we can't do that kind of changes that relies on such internals of other code (if that can be avoided). So I'm against it.

Reply:
Thanks @mlewand for sharing your views. I fully agree, the first one is not as efficient as it could be and the third one is smart (or rather hacky), tricky and subject to some internal changes. Thus **we should go with the second solution**.